src/osd/OSD.cc

```diff
@@ -8308,63 +8311,72 @@ void OSD::handle_osd_map(MOSDMap *m)
+    }
+    if (!(previous_lastmap = service.try_get_map(superblock.get_newest_map()))) {
+      // this is unexpected, either fallback to the first added map:
+      // previous_lastmap = added_maps[first];
+      // or abort:
+      ceph_abort();
     }
-    ceph_assert(lastmap->get_epoch() + 1 == added_map->get_epoch());
-    for (auto& [pool_id, pg_pool] : lastmap->get_pools()) {
-      if (!added_map->have_pg_pool(pool_id)) {
-	pg_num_history.log_pool_delete(added_map_epoch, pool_id);
-	dout(10) << __func__ << " recording final pg_pool_t for pool "
-		 << pool_id << dendl;
-	// this information is needed by _make_pg() if have to restart before
-	// the pool is deleted and need to instantiate a new (zombie) PG[Pool].
-	ghobject_t obj = make_final_pool_info_oid(pool_id);
-	bufferlist bl;
-	encode(pg_pool, bl, CEPH_FEATURES_ALL);
-	string name = lastmap->get_pool_name(pool_id);
-	encode(name, bl);
-	map<string,string> profile;
-	if (lastmap->get_pg_pool(pool_id)->is_erasure()) {
-	  profile = lastmap->get_erasure_code_profile(
-	    lastmap->get_pg_pool(pool_id)->erasure_code_profile);
-	}
-	encode(profile, bl);
-	t.write(coll_t::meta(), obj, 0, bl.length(), bl);
-      } else if (unsigned new_pg_num = added_map->get_pg_num(pool_id);
-		 new_pg_num != pg_pool.get_pg_num()) {
-	dout(10) << __func__ << " recording pool " << pool_id << " pg_num "
-		 << pg_pool.get_pg_num() << " -> " << new_pg_num << dendl;
-	pg_num_history.log_pg_num_change(added_map_epoch, pool_id, new_pg_num);
+  }
+
+  // 1) Check if a pool was deleted
+  for (auto& [pool_id, pg_pool] : previous_lastmap->get_pools()) {
+    if (!current_lastmap->have_pg_pool(pool_id)) {
+      pg_num_history.log_pool_delete(last, pool_id);
+      dout(10) << __func__ << " recording final pg_pool_t for pool "
+               << pool_id << dendl;
+      // this information is needed by _make_pg() if have to restart before
+      // the pool is deleted and need to instantiate a new (zombie) PG[Pool].
+      ghobject_t obj = make_final_pool_info_oid(pool_id);
+      bufferlist bl;
+      encode(pg_pool, bl, CEPH_FEATURES_ALL);
+      string name = previous_lastmap->get_pool_name(pool_id);
+      encode(name, bl);
+      map<string,string> profile;
+      if (previous_lastmap->get_pg_pool(pool_id)->is_erasure()) {
+        profile = previous_lastmap->get_erasure_code_profile(
+        previous_lastmap->get_pg_pool(pool_id)->erasure_code_profile);
       }
+      encode(profile, bl);
+      t.write(coll_t::meta(), obj, 0, bl.length(), bl);
+
+    // 2) For existing pools, check if pg_num was changed
+    } else if (unsigned new_pg_num = current_lastmap->get_pg_num(pool_id);
+               new_pg_num != pg_pool.get_pg_num()) {
+      dout(10) << __func__ << " recording pool " << pool_id << " pg_num "
+               << pg_pool.get_pg_num() << " -> " << new_pg_num << dendl;
+      pg_num_history.log_pg_num_change(last, pool_id, new_pg_num);
```

Comment:
Same question here, last probably isn't the epoch that actually switched to new_pg_num.